File: 0016-3sum-closest/0016-3sum-closest.py
"""
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        closest = nums[0] + nums[1] + nums[2]
        min_diff = abs(target - closest)
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                for k in range(j+1, len(nums)):
                    this = nums[i] + nums[j] + nums[k]
                    if abs(target - this) < min_diff:
                        min_diff = abs(target - this)
                        closest = this
        return closest                
"""

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest = nums[0] + nums[1] + nums[2]
        for i in range(len(nums)-2):
            left, right = i+1, len(nums)-1
            while left < right:
                this = nums[i] + nums[left] + nums[right]
                if this == target:
                    return this
                if abs(target - this) < abs(target - closest):
                    closest = this
                    
                if this < target:
                    left += 1
                else:
                    right -= 1
        return closest            

File: 0016-3sum-closest/0016-3sum-closest.go
import (
    "sort"
)

func threeSumClosest(nums []int, target int) int {
    sort.Ints(nums)
    closest := nums[0] + nums[1] + nums[2]
    for i := 0; i < len(nums)-2; i++ {
        left, right := i+1, len(nums)-1
        for left < right {
            sum := nums[i] + nums[left] + nums[right]
            if sum == target {
                return sum
            }
            if abs(target - sum) < abs(target - closest) {
                closest = sum
            }
            
            if sum < target {
                left++
            } else {
                right--
            }
        }
    }
    return closest
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}